### PR TITLE
integration: Add test to verify behaviour with offline backends

### PR DESCRIPTION
**Pull request by akshaymankar**
Tracked by https://wearezeta.atlassian.net/browse/WPB-2890.

## Checklist

 - [x] ~Add a new entry in an appropriate subdirectory of `changelog.d`~ No changelog
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: hack/helmfile.yaml [hunk 1]
```diff
@@ -129,6 +129,12 @@ releases:
         value: {{ .Values.federationDomain1 }}
       - name: brig.config.optSettings.setFederationDomainConfigs[0].domain
         value: {{ .Values.federationDomain2 }}
+      - name: brig.config.optSettings.setFederationDomainConfigs[2].domain
+        value: {{ .Values.dynBackendDomain1 }}
+      - name: brig.config.optSettings.setFederationDomainConfigs[3].domain
+        value: {{ .Values.dynBackendDomain2 }}
+      - name: brig.config.optSettings.setFederationDomainConfigs[4].domain
+        value: {{ .Values.dynBackendDomain3 }}
```
Comment (mdimjasevic) [inline hack/helmfile.yaml:137]: I don't understand this, but two lines above the index was 0, and this block starts with 2. Is index 1 skipped on purpose?

File: hack/helmfile.yaml [hunk 2]
```diff
@@ -147,5 +153,11 @@ releases:
         value: {{ .Values.federationDomain2 }}
       - name: brig.config.optSettings.setFederationDomainConfigs[0].domain
         value: {{ .Values.federationDomain1 }}
+      - name: brig.config.optSettings.setFederationDomainConfigs[2].domain
+        value: {{ .Values.dynBackendDomain1 }}
+      - name: brig.config.optSettings.setFederationDomainConfigs[3].domain
+        value: {{ .Values.dynBackendDomain2 }}
+      - name: brig.config.optSettings.setFederationDomainConfigs[4].domain
+        value: {{ .Values.dynBackendDomain3 }}
```
Comment (mdimjasevic) [inline hack/helmfile.yaml:161]: ditto.

File: integration/test/Test/Federation.hs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+isNotifForUser :: (MakesValue user, MakesValue a) => user -> a -> App Bool
+isNotifForUser user n = fieldEquals n "payload.0.data.qualified_user_ids.0" (objQidObject user)
+
+fieldEquals :: (MakesValue a, MakesValue b) => a -> String -> b -> App Bool
+fieldEquals a fieldSelector b = do
+  ma <- lookupField a fieldSelector `catchAll` const (pure Nothing)
+  case ma of
+    Nothing -> pure False
+    Just f ->
+      f `isEqual` b
```
Comment (mdimjasevic) [inline integration/test/Test/Federation.hs:164]: These general utilities should be in a separate module, shouldn't they?

**Review by mdimjasevic — COMMENTED** [review]
I'm yet to wrap my head around the new test.

Some comments for the rest are inlined.

**Comment by mdimjasevic** [comment]
I managed to get this test to fail locally once:

```
----- Federation.testNotificationsForOfflineBackends FAIL (4.02 s) -----                                                                                                                      
assertion failure:                                                                                                                                                                            
Expected 1 matching events, but got 0.                                                                                                                                                        
Details:                                                                                                                                                                                      AwaitResult                                                                                                                                                                                   
    success: False                                                                                                                                                                                matches:                                                                                                                                                                                  
                                                                                                                                                                                              
    non-matches:                                                                                                                                                                              
    {                                                                                                                                                                                         
        "id": "d342d3df-4002-11ee-8001-960000bf3df4",                                                                                                                                         
        "payload": [                                                                                                                                                                                      {                                                                                                                                                                                 
                "conversation": "3ecb40c6-d3cd-489f-af4d-c55b11e40722",                                                                                                                                       "data": {                                                                                                                                                                     
                    "user_ids": [                                                                                                                                                             
                        "97a6eafc-49d4-425c-9322-488ae218cb01"                                                                                                                                
                    ],                                                                                                                                                                        
                    "users": [                                                                                                                                                                
                        {                                                                                                                                                                                                 "conversation_role": "wire_admin",                                                                                                                                
                            "id": "97a6eafc-49d4-425c-9322-488ae218cb01",                                                                                                                     
                            "qualified_id": {                                                                                                                                                 
                                "domain": "b.example.com",                                                                                                                                    
                                "id": "97a6eafc-49d4-425c-9322-488ae218cb01"                                                                                                                  
                            }                                                                                                                                                                                         }                                                                                                                                                                     
                    ]                                                                                                                                                                         
                },                                                                                                                                                                            
                "from": "9acfd557-c55e-43a3-9285-ef558d1c7a59",                                                                                                                               
                "qualified_conversation": {                                                                                                                                                   
                    "domain": "example.com",                                                                                                                                                                      "id": "3ecb40c6-d3cd-489f-af4d-c55b11e40722"                                                                                                                              
                },                                                                                                                                                                            
                "qualified_from": {                                                                                                                                                           
                    "domain": "example.com",                                                                                                                                                  
                    "id": "9acfd557-c55e-43a3-9285-ef558d1c7a59"                                                                                                                              
                },                                                                                                                                                                                            "time": "2023-08-21T09:12:17.347Z",                                                                                                                                           
                "type": "conversation.member-join"                                                                                                                                                        }                                                                                                                                                                                 
        ]                                                                                                                                                                                     
    }                                                                                                                                                                                         
    {                                                                                                                                                                                         
        "id": "d3458a9a-4002-11ee-8001-960000bf3df4",                                                                                                                                         
        "payload": [                                                                                                                                                                [178/1958]
            {
                "conversation": "3ecb40c6-d3cd-489f-af4d-c55b11e40722",
                "data": {
                    "qualified_user_ids": [
                        {
                            "domain": "b.example.com",
                            "id": "97a6eafc-49d4-425c-9322-488ae218cb01"
                        }
                    ],
                    "user_ids": [
                        "97a6eafc-49d4-425c-9322-488ae218cb01"
                    ]
                },
                "from": "9acfd557-c55e-43a3-9285-ef558d1c7a59",
                "qualified_conversation": {
                    "domain": "example.com",
                    "id": "3ecb40c6-d3cd-489f-af4d-c55b11e40722"
                },
                "qualified_from": {
                    "domain": "example.com",
                    "id": "9acfd557-c55e-43a3-9285-ef558d1c7a59"
                },
                "time": "2023-08-21T09:12:17.367Z",
                "type": "conversation.member-leave"
            }
        ]
    }
    {
        "id": "d35032d6-4002-11ee-8001-960000bf3df4",
        "payload": [
            {
                "conversation": "3ecb40c6-d3cd-489f-af4d-c55b11e40722",
                "data": {
                    "qualified_user_ids": [
                        {
                            "domain": "example.com",
                            "id": "9acfd557-c55e-43a3-9285-ef558d1c7a59"
                        }
                    ],
                    "user_ids": [
                        "9acfd557-c55e-43a3-9285-ef558d1c7a59"
                    ]
                },
                "from": "9acfd557-c55e-43a3-9285-ef558d1c7a59",
                "qualified_conversation": {
                "qualified_conversation": {
                    "domain": "example.com",
                    "id": "3ecb40c6-d3cd-489f-af4d-c55b11e40722"
                },
                "qualified_from": {
                    "domain": "example.com",
                    "id": "9acfd557-c55e-43a3-9285-ef558d1c7a59"
                },
                "time": "2023-08-21T09:12:17.436Z",
                "type": "conversation.member-leave"
            }
        ]
    }
```

**Comment by akshaymankar** [comment]
@mdimjasevic Can you also post the stack trace for the failing test, it is very hard to tell just from the failure because almost all assertions are about expecting some notification.

**Comment by mdimjasevic** [comment]
> @mdimjasevic Can you also post the stack trace for the failing test, it is very hard to tell just from the failure because almost all assertions are about expecting some notification.

Ah, my bad. I've added it to the original report.

**Review by mdimjasevic — APPROVED** [review]

File: integration/test/Test/Federation.hs (hunk 3 above)
Comment (akshaymankar) [reply integration/test/Test/Federation.hs]: Good idea, moved.

**Comment by mdimjasevic** [comment]
There was a flaky test failure. I've retriggered the CI.

----- Brig.testRemoteUserSearch FAIL (5.35 s) -----

assertion failure:

Actual:

533

Expected:

201


call stack: 

1. assertFailure at test/Testlib/Assertions.hs:52


2. shouldMatch at test/Testlib/Assertions.hs:98


3. shouldMatchInt at test/SetupHelpers.hs:52


4. connectUsers at test/SetupHelpers.hs:63


5. createAndConnectUsers at test/Test/Brig.hs:191


6. k at test/Testlib/ModService.hs:151


7. startDynamicBackends at test/Test/Brig.hs:188




request: 

POST http://127.0.0.1:35161/v4/connections/dynamic-backend-2.test-7ubmn9qetcap.svc.cluster.local/89fd2029-8917-454a-968e-8fd1ec21bd1b

request headers: 

Z-Connection: conn

Z-User: 8f3cbdd6-b5b7-4faa-9ce4-a640a3f816a3

request body:


response status: 533

response body:

{

    "code": 533,

    "data": {

        "domain": "dynamic-backend-2.test-7ubmn9qetcap.svc.cluster.local",

        "path": "/federation/api-version",

        "type": "federation"

    },

    "label": "federation-remote-error",

    "message": "A remote federator failed with status code 503"

}

Went away in the new CI run.

**Pull request merged**